## Calibration information

# Intrinsics
cam_model: Pinhole
cam_height: 375
cam_width: 1242

# kitti 05
cam_fx: 707.0912
cam_fy: 707.0912
cam_cx: 601.8873
cam_cy: 183.1104

cam_d0: 0.0
cam_d1: 0.0
cam_d2: 0.0
cam_d3: 0.0

### Extrinsics between robot body (frame "I") and camera ("C")

I_p_C: [0.0, 0.0, 0.0]
I_q_C: [-0.5, 0.5, -0.5, 0.5] # IMU x forward, y left, z up <-> camera x right, y down, z forward

## Optimization parameters
# covariance_delay: 5.0
covariance_delay: 0.1
max_optimization_time: 0.5
loop_closure_threshold: 50 # in number of keyframes
smoothing_length: 100

verbose_optimization: true
use_manual_elimination_ordering: false


## Feature_tracker parameters
ransac_iterations: 35       # number of iterations in essential matrix estimating ransac loop
feature_spacing: 10         # minimum pixel distance between extracted features. i.e. an upper bound of sorts on feature density
max_features_per_im: 500    # maximum number of features to keep per image
sqrt_samp_thresh: 0.01      # Threshold on inliers in sampson error, higher = more permissive
tracking_framerate: 10     # drops images so tracking is performed at this framerate

## Semslam parameters

odometry_type: "external"   # odometry is handled by the VISO node, so "external"

### Geometric feature parameters
include_geometric_features: true
reprojection_error_threshold: 4
use_smart_projection_factors: true

### Semantic measurement parameters
keypoint_msmt_sigma: 12                 # keypoint measurement sigma (pixels)
keypoint_initialization_depth_sigma: 2  # meters

min_object_n_keypoints: 5               # minimum number of observed keypoints needed to consider an object good and include it in the factor graph
min_landmark_observations: 3           # minimum number of observations needed for a keypoint before estimating depth and including in optimization
min_observed_keypoints_to_initialize: 4

keyframe_translation_threshold: 1
keyframe_translation_without_measurement_threshold: 2.5
keyframe_rotation_threshold: 10 # degrees
keyframe_rotation_without_measurement_threshold: 15

### Semantic object optimization parameters
structure_regularization_factor: 5     # lambda in structure optimization term lambda * ||c||^2
robust_estimator_parameter: 1.5
include_objects_in_graph: true

# Error in the structure factor due to misalignment is defined as 
# || L - R*S - t ||_W
# where W = (1/structure_error_coefficient) * I
# i.e. this is an inverse "noise" like value where higher = more penalty for 
# deviations from the calculated object structure
structure_error_coefficient: 100

keypoint_activation_threshold: 0.2     # threshold above which to consider a keypoint observed

# camera_range: 10                        # meters

mahal_thresh_assign: 4                  # Mahalanobis distance below which to assign data (MLDataAssociator)
mahal_thresh_init: 12                   # Mahalanobis distance above which to assign new mapped landmarks (MLDataAssociator)

# max_new_factor_error: 300

constraint_weight_threshold: 0.8        # if the weight of a constraint (i.e., an object measurement) is less than this do not incorporate the keypoint measurements corresponding to this object
new_landmark_weight_threshold: 0.9      # if the weight of the last column in the weights matrix for a measurement is greater than this value, add a new object to the estimated_objects_ vector

## Object classes
model_names: ["car"]
model_directory: "/home/kodlab/catkin_ws/src/object_pose_detection/semslam/semslam/models/objects/"

gravity: [0, 0, -9.81]

# init_p: [0, 0, 0]
# init_p_sigma: [1.0e-7, 1.0e-7, 1.0e-7]

# init_q: [0, 0, 0, 1]
# init_q_sigma: [0.02, 0.02, 0.02]

## Visualization
geometric_point_scale: 0.4
object_keypoint_scale: 0.5
show_object_labels: true
trajectory_width: 0.25
covisibility_width: 0.01
draw_semantic_covisibility: true
draw_geometric_covisibility: false
color_smoothing_window: true